ocaml/xapi/xapi_host.ml

```diff
@@ -598,46 +612,93 @@ let evacuate ~__context ~host ~network =
     plans ;
   (* Do it *)
   let individual_progress = 1.0 /. float (Hashtbl.length plans) in
-  let migrate_vm vm plan =
-    match plan with
-    | Migrate host ->
-        ( try
-            ( if network <> Ref.null then
-                let hosts = Db.Host.get_all ~__context in
-                List.iter
-                  (fun host ->
-                    ignore
-                    @@ Xapi_network_attach_helpers
-                       .assert_valid_ip_configuration_on_network_for_host
-                         ~__context ~self:network ~host
-                  )
-                  hosts
-            ) ;
-            let with_network_option =
-              if network <> Ref.null then
-                [("network", Ref.string_of network)]
-              else
-                []
-            in
-            let options = ("live", "true") :: with_network_option in
-            Helpers.call_api_functions ~__context (fun rpc session_id ->
-                Client.Client.VM.pool_migrate ~rpc ~session_id ~vm ~host
-                  ~options
-            )
-          with
-        | Api_errors.Server_error (code, _)
-          when code = Api_errors.vm_bad_power_state ->
+  let migrate_vm ~rpc ~session_id (vm, plan) =
+    let sub_task =
+      let options = [("live", "true")] in
+      match plan with
+      | Migrate host when network = Ref.null ->
+          Client.Client.Async.VM.pool_migrate ~rpc ~session_id ~vm ~host
+            ~options
+      | Migrate host ->
+          let options = ("network", Ref.string_of network) :: options in
+          Db.Host.get_all ~__context
+          |> List.iter (fun host ->
+                 ignore
+                 @@ Xapi_network_attach_helpers
+                    .assert_valid_ip_configuration_on_network_for_host
+                      ~__context ~self:network ~host
+             ) ;
+          Client.Client.Async.VM.pool_migrate ~rpc ~session_id ~vm ~host
+            ~options
+      | Error (code, params) ->
+          (* should never happen *)
+          raise (Api_errors.Server_error (code, params))
+    in
+    let progress = Db.Task.get_progress ~__context ~self:task in
+    TaskHelper.set_progress ~__context (progress +. individual_progress) ;
```

Comment:
This now assigns progress immediately after the async call returns rather than at the end of the migration.

Reply:
That's true. Maybe progress it after each batch has completed?